File: day3/part2.py
import re

def readfile(filename):
    with open(filename) as file:
        return [line.rstrip('\n') for line in file]

def extract(line):
    return map(int, re.match(r'#(\d+)\s*@\s*(\d+),(\d+):\s*(\d+)x(\d+)', line).groups())

def main():
    n = 1000
    fabric = [[[] for _ in range(n)] for _ in range(n)]
    all_ids = []

    for line in readfile('input'):
        identifier, x, y, w, h = extract(line)
        all_ids.append(identifier)
        for i in range(w):
            for j in range(h):
                fabric[x + i][y + j].append(identifier)

    for i in range(n):
        for j in range(n):
            if len(fabric[i][j]) > 1:
                for identifier in fabric[i][j]:
                    if identifier in all_ids:
                        all_ids.remove(identifier)

    print(all_ids)

main()

File: day07/part2.go
package main

import (
"bufio" 
"fmt" 
"os"
"strings"
)

type Dependency struct {
	from, to byte
}

type Work struct {
	time, item int
}

func readFile(filename string) []Dependency {
	file, _ := os.Open(filename)
	defer file.Close()

	var deps []Dependency
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		from, end := 5, 36
		line := scanner.Text()
		deps = append(deps, Dependency{line[from] - 'A', line[end] - 'A'})
	}

	return deps;
}

func createTasks(deps []Dependency) [26]string {
	var tasks [26]string
	for i := 0; i < 26; i++ {
		tasks[i] = "?"
	}

	for _, dep := range deps {
		newDep := string('A' + dep.from)

		if tasks[dep.from] == "?" {
			tasks[dep.from] = ""
		}
		if tasks[dep.to] == "?" {
			tasks[dep.to] = ""
		}

		tasks[dep.to] = tasks[dep.to] + newDep
	}

	return tasks
}

func findWork(tasks [26]string) int {
	for i, task := range tasks {
		if task == "" {
			return i;
		}
	}
	return -1;
}


func run(nworkers int, timePenalty int, tasks [26]string) int {
	workers := make([]Work, nworkers)
	for i := 0; i < nworkers; i++ {
		workers[i] = Work{0, 0}
	}

	time := -1
	for hasWorked := true; hasWorked; time++ {
		hasWorked = false

		for i, worker := range workers {
			if worker.time == 0 {
				item := findWork(tasks)
				if item != -1 {
					hasWorked = true
					workers[i] = Work{timePenalty + item + 1, item}
					tasks[item] = "?"
				}
			}
		}

		for i, worker := range workers {
			if worker.time > 0 {
				workers[i].time--
				hasWorked = true

				if workers[i].time == 0 {
					for j := 0; j < 26; j++ {
						tasks[j] = strings.Replace(tasks[j], string('A' + workers[i].item), "", -1)
					}
				}
			}
		}
	}
	return time
}

func main() {
	tasks := createTasks(readFile("input"))

	fmt.Println(run(5, 60, tasks))
}